### PR TITLE
[mxfp8 moe training] add triton kernel for mxfp8 quantization along dim0

### PR DESCRIPTION
**Pull request by danielvegamyhre**
[mxfp8 moe training] add triton kernel for mxfp8 quantization along dim0

### CODE REVIEW

File: torchao/prototype/mx_formats/kernels.py [hunk 1]
```diff
@@ -1038,6 +1040,175 @@ def to_mxfp8_dim1_kernel(
...
 
+    @triton.autotune(
+        configs=_get_mxfp8_dim1_kernel_autotune_configs(),
+        key=["n_rows", "n_cols", "INNER_BLOCK_SIZE"],
+    )
+    @triton.jit
+    def to_mxfp8_dim0_kernel(
+        x_ptr,  # pointer to input tensor
+        output_ptr,  # pointer to output tensor (row-normalized)
+        row_scale_ptr,  # pointer to store row-wise maximum absolute values
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/kernels.py:1049]: should the comment say scale or max_abs? also row-wise should really say "across dim0"?

File: torchao/prototype/mx_formats/kernels.py [hunk 2]
```diff
@@ -1038,6 +1040,175 @@ def to_mxfp8_dim1_kernel(
...
+        jump_vals_per_row = (n_cols - COL_TILE_SIZE) // INNER_BLOCK_SIZE
+
+        # example transformation (specifics depend on tile sizes):
+        # [0, 1, 2, 3, 4, 5, 6, 7] -> [0, 1, 4, 5, 8, 9, 12, 13]
+        row_scale_indices = row_scale_indices + (
+            (row_scale_indices // BLOCKS_PER_COL_TILE) * jump_vals_per_row
+        )
+
+        # Store the scales
+        tl.store(row_scale_start_ptr + row_scale_indices, row_scale_e8m0)
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/kernels.py]: should this either have a mask or a TODO to add masking?

**Review by vkuzo — APPROVED** [review]

File: torchao/prototype/mx_formats/kernels.py (hunk 2 above)
Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/kernels.py]: I added masking in https://github.com/pytorch/ao/pull/3131 where I refactored/simplified the kernel, but I'll go ahead and move it here

File: torchao/prototype/mx_formats/kernels.py (hunk 1 above)
Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/kernels.py:1049]: the dim1 kernel said "column-wise" here, so i just changed it to say "row-wise" here for consistency. dim0/dim1 is also fine. The comment shouldn't "maximum absolute values" for sure though, I will remove that from both kernels.

File: torchao/prototype/mx_formats/kernels.py [hunk 3]
```diff
@@ -829,6 +829,8 @@ def _(uint8_data):
     import triton.language as tl
     from torch.library import triton_op, wrap_triton
 
+    print("importing triton ops")
```
Comment (drisspg) [inline torchao/prototype/mx_formats/kernels.py]: remove please

File: torchao/prototype/mx_formats/kernels.py [hunk 4]
```diff
@@ -1038,6 +1040,175 @@ def to_mxfp8_dim1_kernel(
...
+        jump_vals_per_row = (n_cols - COL_TILE_SIZE) // INNER_BLOCK_SIZE
+
+        # example transformation (specifics depend on tile sizes):
+        # [0, 1, 2, 3, 4, 5, 6, 7] -> [0, 1, 4, 5, 8, 9, 12, 13]
+        row_scale_indices = row_scale_indices + (
+            (row_scale_indices // BLOCKS_PER_COL_TILE) * jump_vals_per_row
+        )
+
+        # Store the scales
+        tl.store(row_scale_start_ptr + row_scale_indices, row_scale_e8m0)
```
Comment (drisspg) [inline torchao/prototype/mx_formats/kernels.py]: is this storing in swizzled layout?

Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/kernels.py]: No, we have separate kernels for that. Low hanging fruit could be to write directly to blocked swizzled format here, like Cursor mentioned they're doing in this blog https://cursor.com/blog/kernels

File: torchao/prototype/mx_formats/kernels.py (hunk 3 above)
Comment (danielvegamyhre) [reply torchao/prototype/mx_formats/kernels.py]: Woops, was debugging an import error, thanks

File: torchao/prototype/mx_formats/kernels.py (hunk 4 above)
Comment (drisspg) [reply torchao/prototype/mx_formats/kernels.py]: Its a non zero perf win for activation casting in inference. We can leave as a follow up but should be pretty easy to just have aconstexpr path do this
